File: leetcode/easy/589-nary-preorder/recursive.py
"""
  @author Vortexx2
  Problem 589 - N-ary Preorder Traversal

  Runtime - 52 ms
  Memory Usage - 15.9 MB
"""
from typing import List


# Definition for a Node.
class Node:
  def __init__(self, val=None, children=None):
    self.val = val
    self.children = children


class Solution:
  def preorder(self, root: 'Node') -> List[int]:
    res = []
    self.helper(root, res)

    return res

  def helper(self, root: 'Node', arr: List[int]) -> None:

    if root:
      arr.append(root.val)

      for child in root.children:
        self.helper(child, arr)

File: leetcode/easy/589-nary-preorder/recursive.cpp
/**
 * @file recursive.cpp
 * @author Vortexx2
 * @brief Problem 589 - N-ary Preorder Traversal
 * @date 20-07-2021
 * 
 * Runtime - 20 ms
 * Memory Usage - 11.3 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

class Node
{
public:
  int val;
  vector<Node *> children;

  Node() {}

  Node(int _val)
  {
    val = _val;
  }

  Node(int _val, vector<Node *> _children)
  {
    val = _val;
    children = _children;
  }
};

class Solution
{
public:
  vector<int> preorder(Node *root)
  {
    vector<int> res;
    helper(root, res);

    return res;
  }

  void helper(Node *root, vector<int> &arr)
  {
    if (root)
    {
      arr.push_back(root->val);
      for (auto &child : root->children)
      {
        helper(child, arr);
      }
    }
  }
};